asab/library/providers/git.py

```diff
@@ -0,0 +1,145 @@
+import os
+import tempfile
+import logging
+import hashlib
+
+import pygit2
+
+from .filesystem import FileSystemLibraryProvider
+from ...config import Config
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class GitLibraryProvider(FileSystemLibraryProvider):
+	"""
+	Read-only git provider to read from remote repository.
+	It clones a remote git repository to a temporary directory and then uses the
+	FileSystemLibraryProvider to read the files.
+	To read from local git repository, please use FileSystemProvider.
+
+	Configuration:
+	(Use either deploytoken, publickey+privatekey for SSH option, or username and password and HTTP access.)
+
+
+	```
+	[library]
+	providers=git+<URL or deploy token>
+
+	[library:git]
+	publickey=<absolute path to file>
+	privatekey=<absolute path to file>
+	username=johnsmith
+	password=secretpassword
+	repodir=<optional location of the repository cache>
+	```
+	"""
+	def __init__(self, library, path):
+		self.LastCommit = None
+		self.URL = path[4:]
+
+		self.Callbacks = pygit2.RemoteCallbacks(get_git_credentials(self.URL))
+
+		# TODO: Check `repodir`
+		tempdir = tempfile.gettempdir()
+		self.RepoPath = os.path.join(
+			tempdir,
+			"asab.library.git",
+			hashlib.sha256(self.URL.encode('utf-8')).hexdigest()
+		)
+
+		try:
+			# Clone a new repository
+			os.makedirs(self.RepoPath, mode=0o700)
+			self.GitRepository = pygit2.clone_repository(self.URL, self.RepoPath, callbacks=self.Callbacks)
+			self._check_remote()
+
+		except FileExistsError:
+			# Update the existing repository
+			self.GitRepository = pygit2.Repository(self.RepoPath)
+			self._check_remote()
+			commit_id = fetch(self.GitRepository, self.Callbacks)
+			merge(self.GitRepository, commit_id)
+
+		super().__init__(library, self.RepoPath)
+
+		from ...proactor import Module
+		self.App.add_module(Module)
+		self.ProactorService = self.App.get_service("asab.ProactorService")
+
+		self.App.PubSub.subscribe("Application.tick/60!", self._periodic_pull)
+
+
+	async def pull(self):
+		"""
+		Equivalent to `git pull` command.
+		"""
+		commit_id = await self.ProactorService.execute(fetch, self.GitRepository, self.Callbacks)
+		if commit_id == self.LastCommit:
+			return
+		self.LastCommit = commit_id
+		await self.ProactorService.execute(merge, self.GitRepository, commit_id)
+
+
+	async def _periodic_pull(self, event_name):
+		await self.pull()
+		self.App.PubSub.publish("GitLibraryProvider.pull!")
```

Comment:
That's not needed I guess - but the detection if there is a change is welcomed here.
It will plug into an automated library reload on change.